checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1605 @@
+   * be because the method has no return type or because it is annotated with {@link NotOwning}.
+   *
+   * @param facts the current set of facts
+   * @param node the invocation node to check
+   * @return true iff the result of node should not be tracked in facts
+   */
+  private boolean skipTrackingInvocationResult(
+      Set<ImmutableSet<LocalVarWithTree>> facts, Node node) {
+    Tree callTree = node.getTree();
+    if (callTree.getKind() == Tree.Kind.METHOD_INVOCATION) {
+      MethodInvocationTree methodInvokeTree = (MethodInvocationTree) callTree;
+
+      if (TreeUtils.isSuperConstructorCall(methodInvokeTree)
+          || TreeUtils.isThisConstructorCall(methodInvokeTree)) {
+        handleThisOrSuperConstructorMustCallAlias(facts, node);
+        return true;
+      }
+      return returnTypeIsMustCallAliasWithIgnorable((MethodInvocationNode) node)
+          || hasNotOwningReturnType((MethodInvocationNode) node);
+    }
+    return false;
+  }
+
+  /**
+   * Returns true if this node represents a method invocation of a must-call-alias method, where the
+   * other must call alias is some ignorable pointer, such as an owning field or a pointer that is
+   * guaranteed to be non-owning, such as "`this`" or a non-owning field.
+   *
+   * @param node a method invocation node
+   * @return true if this is the invocation of a method whose return type is MCA with an owning
+   *     field or a non-owning pointer
+   */
+  private boolean returnTypeIsMustCallAliasWithIgnorable(MethodInvocationNode node) {
+    Node mcaParam = getMustCallAliasParamVar(node);
+    return mcaParam instanceof FieldAccessNode || mcaParam instanceof ThisNode;
+  }
+
+  /**
+   * Checks if {@code node} is nested inside a {@link TypeCastNode} or a {@link
+   * TernaryExpressionNode}, by looking at the successor block in the CFG.
+   *
+   * @param node the CFG node
+   * @return {@code true} if {@code node} is in a {@link SingleSuccessorBlock} {@code b}, the first
+   *     {@link Node} in {@code b}'s successor block is a {@link TypeCastNode} or a {@link
+   *     TernaryExpressionNode}, and {@code node} is an operand of the successor node; {@code false}
+   *     otherwise
+   */
+  private boolean nestedInCastOrTernary(Node node) {
+    if (!(node.getBlock() instanceof SingleSuccessorBlock)) {
+      return false;
+    }
+    Block successorBlock = ((SingleSuccessorBlock) node.getBlock()).getSuccessor();
+    if (successorBlock != null) {
+      List<Node> succNodes = successorBlock.getNodes();
+      if (succNodes.size() > 0) {
+        Node succNode = succNodes.get(0);
+        if (succNode instanceof TypeCastNode) {
+          return ((TypeCastNode) succNode).getOperand().equals(node);
+        } else if (succNode instanceof TernaryExpressionNode) {
+          TernaryExpressionNode ternaryExpressionNode = (TernaryExpressionNode) succNode;
+          return ternaryExpressionNode.getThenOperand().equals(node)
```

Comment:
To convey the concept (in addition to the implementation details), I think it would be helpful to give intuition, such as that these are all the "pass-through" or "no-op" operators.

Reply:
@msridhar

Reply:
558f581f74129722220f3dde7f94784a4b637569